src/Endpoint/EventsEndpoint.php

```diff
@@ -3,70 +3,122 @@
     }
 
     /**
-     * @param string $verb
-     * @param string $ip
-     * @param array $user
-     * @param string|null $userAgent
+     * Tracks an event using the ThisData API.
+     * @param string $verb           Describes what the User did, using an English present tense verb
+     * @param string $ip             The IP address of the client logging in
+     * @param array|null $user       An optional array containing id, and optionally name, email, mobile.
+     * @param string|null $userAgent The browser user agent of the client logging in
+     * @param array|null $source     Source details (e.g. for multi-tenanted applications)
+     * @param array|null $session    Extra information that provides useful context about the session, for example the session ID, or some cookie information
+     * @param array|null $device     Information about the device being used
+     * @see AbstractEndpoint         Predefined VERB_ constants
      */
-    public function trackEvent($verb, $ip, array $user, $userAgent = null)
+    public function trackEvent($verb, $ip, array $user = null, $userAgent = null, array $source = null, array $session = null, array $device = null)
     {
-        $this->execute('POST', ThisData::ENDPOINT_EVENTS, [
+        $event = array(
             self::PARAM_VERB => $verb,
-            self::PARAM_IP   => $ip,
-            self::PARAM_USER => [
+            self::PARAM_IP   => $ip
+        );
+
+        if (!is_null($userAgent)) {
+            $event[self::PARAM_USER_AGENT] = $userAgent;
+        }
+        if (!is_null($user)) {
+            $event[self::PARAM_USER] = array_filter([
                 self::PARAM_USER__ID     => $this->findValue(self::PARAM_USER__ID, $user),
                 self::PARAM_USER__NAME   => $this->findValue(self::PARAM_USER__NAME, $user),
                 self::PARAM_USER__EMAIL  => $this->findValue(self::PARAM_USER__EMAIL, $user),
                 self::PARAM_USER__MOBILE => $this->findValue(self::PARAM_USER__MOBILE, $user),
-            ],
-            self::PARAM_USER_AGENT => $userAgent
+                self::PARAM_USER__AUTHENTICATED => $this->findValue(self::PARAM_USER__AUTHENTICATED, $user),
+            ]);
+        }
+        if (!is_null($source)) {
+            $event[self::PARAM_SOURCE] = array_filter([
+                self::PARAM_SOURCE__NAME     => $this->findValue(self::PARAM_SOURCE__NAME, $source),
+                self::PARAM_SOURCE__LOGO_URL => $this->findValue(self::PARAM_SOURCE__LOGO_URL, $source)
+            ]);
+        }
+
+        // Add information about the session
+        // First, the session ID if it's passed
+        $event[self::PARAM_SESSION] = array_filter([
+            self::PARAM_SESSION__ID  => $this->findValue(self::PARAM_SESSION__ID, $session)
         ]);
+        // Then pull the TD cookie if its present
+        if(isset($_COOKIE[ThisData::TD_COOKIE_NAME])) {
```

Comment:
tbh I'm not sure what is idiomatic PHP any more, perhaps Malcolm will have some feedback..